src/fusiontensor/fusiontensor.jl

```diff
@@ -0,0 +1,189 @@
+      length(codomain_legs) + length(domain_legs),
+      typeof(codomain_legs),
+      typeof(domain_legs),
+      typeof(mat),
+      typeof(trees_block_mapping),
+    }(
+      mat, codomain_legs, domain_legs, trees_block_mapping
+    )
+  end
+end
+
+# getters
+data_matrix(ft::FusionTensor) = ft.data_matrix
+codomain_axes(ft::FusionTensor) = ft.codomain_axes
+domain_axes(ft::FusionTensor) = ft.domain_axes
+trees_block_mapping(ft::FusionTensor) = ft.trees_block_mapping
+
+# misc access
+ndims_codomain(ft::FusionTensor) = length(codomain_axes(ft))
+ndims_domain(ft::FusionTensor) = length(domain_axes(ft))
+
+matrix_size(ft::FusionTensor) = quantum_dimension.(axes(data_matrix(ft)))
+matrix_row_axis(ft::FusionTensor) = first(axes(data_matrix(ft)))
+matrix_column_axis(ft::FusionTensor) = last(axes(data_matrix(ft)))
+function charge_block_size(ft::FusionTensor, f1::SectorFusionTree, f2::SectorFusionTree)
+  b = Tuple(findblock(ft, f1, f2))
+  return ntuple(i -> Int(length(axes(ft)[i][b[i]])), ndims(ft))
+end
+
+# GradedUnitRanges interface
+GradedUnitRanges.sector_type(ft::FusionTensor) = sector_type(matrix_row_axis(ft))
+
+# BlockArrays interface
+function BlockArrays.findblock(ft::FusionTensor, f1::SectorFusionTree, f2::SectorFusionTree)
+  # find outer block corresponding to fusion trees
+  @assert ndims_codomain(ft) == length(f1)
+  @assert ndims_domain(ft) == length(f2)
+  @assert sector_type(ft) == sector_type(f1)
+  @assert sector_type(ft) == sector_type(f2)
+  b1 = ntuple(
+    i -> findfirst(==(leaves(f1)[i]), blocklabels(codomain_axes(ft)[i])), ndims_codomain(ft)
+  )
+  b2 = ntuple(
+    i -> findfirst(==(leaves(f2)[i]), blocklabels(dual(domain_axes(ft)[i]))),
+    ndims_domain(ft),
+  )
+  return Block(b1..., b2...)
+end
+
+sanitize_axes(::Tuple{}) = ()
+function sanitize_axes(raw_legs::Tuple{Vararg{AbstractGradedUnitRange}})
+  legs = promote_sectors(typeof(first(raw_legs)), raw_legs)
+  @assert all(check_unique_blocklabels.(legs))
+  return legs
+end
+
+function check_unique_blocklabels(g::AbstractGradedUnitRange)
+  return length(unique(blocklabels(g))) == blocklength(g)
+end
+
+function promote_sectors(
+  ::Type{<:AbstractGradedUnitRange{LA}}, legs::Tuple{Vararg{AbstractGradedUnitRange{LA}}}
+) where {LA}  # nothing to do
+  return legs
+end
+
+function promote_sectors(
+  ::Type{<:AbstractGradedUnitRange}, legs::Tuple{Vararg{AbstractGradedUnitRange}}
+)
+  T = promote_sector_type(legs)
+  # fuse with trivial to insert all missing arguments inside each GradedAxis
+  # avoid depending on SymmetrySectors internals
+  s0 = trivial(T)
+  unified_legs = map_blocklabels.(s -> only(blocklabels(fusion_product(s0, s))), legs)
+  return unified_legs
+end
+
+function promote_sector_type(legs)
+  # fuse trivial sectors to produce unified type
+  # avoid depending on SymmetrySectors internals
+  return sector_type(fusion_product(trivial.(legs)...))
+end
+
+# initialize with already computed data_matrix
+function fusiontensor(
```

Comment:
As discussed, I think we need to decide on a naming convention for these constructors. For now my suggestion would be to call any constructor that takes in a data matrix `FusionTensor` (since those are more literally constructors, i.e. they are doing minimal manipulation of the input data), and any function that is converting from a dense array `to_fusiontensor` (since those are more like converters rather than constructors).

So with that convention, this one should be called `FusionTensor`.